packages/server/src/adapters/zodFileSchema.ts

```diff
@@ -0,0 +1,76 @@
+import type { Readable } from 'stream';
+import z from 'zod';
+
+function isObject(value: unknown): value is Record<string, unknown> {
+  // check that value is object
+  return !!value && !Array.isArray(value) && typeof value === 'object';
+}
+
+const readableEsqueSchema = z.custom<Readable>((value) => {
+  // check if file is is a Readable without importing Readable
+  if (!isObject(value)) {
+    return false;
+  }
+  if (typeof (value as unknown as Readable).on === 'function') {
+    return true;
+  }
+
+  return false;
+});
+
+/**
+ * This is a hack to check if a value is a File in the browser
+ */
+const fileEsqueSchema = z.custom<File>((value) => {
+  if (!isObject(value)) {
+    return false;
+  }
+  if (typeof (value as unknown as File).type === 'string') {
+    return true;
+  }
+  return false;
+});
+
+/**
+ * This is a hack to check if a value is a FileList in the browser
+ * Returns the first item file 🤷‍♂️
+ */
+const fileListEsqueSchema = z
+  .custom<FileList>((value) => {
+    if (!isObject(value)) {
+      return false;
+    }
+    if (
+      typeof (value as unknown as FileList).item === 'function' &&
+      (value as unknown as FileList).item(0)
+    ) {
+      return true;
+    }
+    return false;
+  })
+  .transform((value) => value.item(0) as File);
+
+export const zodFileStreamSchema = z.object({
+  stream: readableEsqueSchema,
+  name: z.string(),
+  type: z.string(),
+});
+
+/**
+ * Isomorphic File schema.
+ * It will accept a File, FileList, or a ReadableStream and return a File
+ */
+export const zodFileSchema = z.union([
+  fileEsqueSchema,
+  fileListEsqueSchema,
+  zodFileStreamSchema.transform(async (input) => {
+    const chunks: Buffer[] = [];
+    for await (const chunk of input.stream) {
+      chunks.push(chunk);
+    }
+
+    return new File(chunks, input.name, {
+      type: input.type,
+    });
+  }),
+]);
```

Comment:
Intent here is to make a zod helper that works for files both in the browser and on the server.

Not sure how to make this work with `.optional()` at the end.

It **works** but it also treats empty files with `.optional()` as an error as the input isn't optional

This is how `ZodOptional`: https://github.com/colinhacks/zod/blob/a1fc3fb815051beb4b7030969828aa6abe469c2c/src/types.ts#L4353

Unsure how to make the `._getType` return undefined to it's actually treated as undefined when

- FileList is empty 
- File is bullish

Does preprocess help here somehow? Idk

Reply:
I don't quite understand what the requirements are here. I also don't know if it's possible for the inferred type of a single schema to differ across the platforms. Here's a version that should work, but the inferred output type will be `FileList | File`. Again, not sure I understood what you need:

```ts

z.custom<FileList | Readable>(
  (input) => {
    if (typeof FileList !== "undefined" && input instanceof FileList)
      return true;
    if (typeof Readable !== "undefined" && input instanceof Readable)
      return true;
    return;
  },
  { message: "Input must be a file." }
).transform((input, ctx) => {
  if (input instanceof FileList) {
    // additional valuation
    if (input.length === 0) {
      ctx.addIssue({
        code: "custom",
        message: "FileList must have at least one file.",
      });
    }
    return input;
  }
  if (input instanceof Readable) {
    // additional validation
    // write to file...?
    return file;
  }
});

```

Reply:
This is what I came up with. I didn't try to fully match the features of what you've already implemented but it might give you some ideeas.

```ts
z.custom<FileList | Readable>(
  (input) => {
    if (typeof FileList !== "undefined" && input instanceof FileList)
      return true;
    if (typeof Readable !== "undefined" && input instanceof Readable)
      return true;
    return;
  },
  { message: "Input must be a file." }
).transform((input, ctx) => {
  if (input instanceof FileList) {
    // additional valuation
    if (input.length === 0) {
      ctx.addIssue({
        code: "custom",
        message: "Please select a file",
      });
    }
  }
  if (input instanceof Readable) {
    // additional validation
    // read file fro stream...
    return new File(...);
  }
});
```